File: scala/P18.scala
// P18 Extract a slice from a list.

// Using "take" and "drop"
def f1[T](list: List[T], i: Int, k: Int): List[T] =
  list.take(k).drop(i - 1)

// Using "splitAt"
def f2[T](list: List[T], i: Int, k: Int): List[T] =
  list.splitAt(k)._1.splitAt(i - 1)._2

// Recursion
def f3[T](list: List[T], i: Int, k: Int): List[T] = {
  list match {
    case Nil => Nil
    case x :: xs =>
      if (i > 1)
        f3(xs, i - 1, k - 1)
      else if (k > 0)
        x :: f3(xs, 1, k - 1)
      else Nil
  }
}

// Tail recursion
@annotation.tailrec
def f4[T](list: List[T], i: Int, k: Int, acc: List[T] = Nil): List[T] = {
  list match {
    case Nil => acc.reverse
    case x :: xs =>
      if (i > 1)
        f4(xs, i - 1, k - 1, acc)
      else if (k > 0)
        f4(xs, 1, k - 1, x :: acc)
      else acc.reverse
  }
}

// Filter by index
def f5[T](list: List[T], i: Int, k: Int): List[T] =
  list.zip(1 to k).filter(_._2 >= i).unzip._1

// Using for comprehension
def f6[T](list: List[T], i: Int, k: Int): List[T] =
  for ((x, n) <- list.zip(1 to k) if n >= i) yield x

// Folding
def f7[T](list: List[T], i: Int, k: Int): List[T] = {
  list.zip(1 to k).foldRight(Nil: List[T]) {
    case ((x, n), acc) => if (n >= i) x :: acc else acc
  }
}

File: haskell/p18.hs
-- P18 Extract a slice from a list.

-- Using "take" and "drop"
f1 :: [a] -> Int -> Int -> [a]
f1 xs i k = drop (pred i) $ take k xs

-- Using "splitAt"
f2 :: [a] -> Int -> Int -> [a]
f2 xs i k = snd $ splitAt (pred i) $ fst $ splitAt k xs

-- Recursion
f3 :: [a] -> Int -> Int -> [a]
f3 [] _ _ = []
f3 (x:xs) i k
    | i > 1 = f3 xs (pred i) (pred k)
    | k > 0 = x : f3 xs 1 (pred k)
    | otherwise = []

-- Tail recursion
f4 :: [a] -> Int -> Int -> [a]
f4 xs i k = reverse $ f [] xs i k
    where f acc [] _ _ = acc
          f acc (x : xs) i k
            | i > 1 = f acc xs (pred i) (pred k)
            | k > 0 = f (x : acc) xs 1 (pred k)
            | otherwise = acc

-- Filter by index
f5 :: [a] -> Int -> Int -> [a]
f5 xs i k = fst $ unzip $ filter ((>=i) . snd) $ zip xs [1..k]

-- List comprehension
f6 :: [a] -> Int -> Int -> [a]
f6 xs i k = [x | (x, n) <- zip xs [1..k], n >= i]

-- Folding
f7 :: [a] -> Int -> Int -> [a]
f7 xs i k = foldr f [] $ zip xs [1..k]
    where f (x, n) acc
            | n >=i = x : acc
            | otherwise = acc